ESQL: Fix lookup join output

alex-spies opened the pull request:
Fix https://github.com/elastic/elasticsearch/issues/117584. More generally, make queries work where the `LOOKUP JOIN`'s `ON` field makes it into the result.

Also relates https://github.com/elastic/elasticsearch/issues/116208:
- ~Make `LOOKUP JOIN` work on data nodes.~ **Argh, the multi node tests still have issues and sometimes return `null`** instead of the correctly returned values. But at least the plans sent to data nodes are consistent with the coordinator's plan.
- Solve the item "Look whether the LookupJoin output could be computed inside the class".

**Note to reviewers:** This is not supposed to be a definitive solution for the output order of `LOOKUP JOIN`. That's still open. Accordingly, the updated tests remain minimal; this PR just aims to enable more queries so we can work on more kinds of query plans without running into bugs.

Currently, the computation of a `LookupJoin` node's output happens in the Analyzer; the computed `List<Attribute>` of output attributes is passed on as part of a new `LookupJoin` node. However, then this is replaced by a plain `Join` node in the surrogate step - which forgets about the computed attributes.
This leads to inconsistencies. It also generally prevents execution of lookup joins on data nodes because they are only ever sent `Join` nodes, not `LookupJoin` (the latter is not serializable and doesn't escape the coordinator). This means that the plain `Join` node **must** be the source of truth when it comes to the output, otherwise things will be inconsistent.

Remove the output computation from the Analyzer and move it into the `Join` node, instead.

alex-spies committed:
Make LookupJoin work on data nodes + fix output

alex-spies committed:
Fix output order

alex-spies committed:
Update capability

alex-spies committed:
Merge branch 'main' into fix-lookup-join-output

alex-spies committed:
Add tests with subsequent EVAL

craigtaverner reviewed (approved):
LGTM - seems mostly to be a simplification!

alex-spies commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/plan/logical/join/Join.java:
```diff
@@ -161,14 +149,6 @@ public static List<Attribute> makeReference(List<Attribute> output) {
         return out;
     }
 
-    private static List<Attribute> makeNullable(List<Attribute> output) {
```
It will come back - but it's been unused for long now, and it really only wraps 1 line.

alex-spies committed:
Fix BinaryPlan.computeReferences

This must not just use the references from its own output. Not only is
this wrong, it also leads to failures when we call the .references()
method on unresolved plans.

alex-spies committed:
Revert testLookup to how it used to be

This reverts the test for the *former* LOOKUP command. Its output is now
more in line again with how it was computed before.

alex-spies commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/plan/logical/BinaryPlan.java:
```diff
@@ -45,11 +43,6 @@ public final BinaryPlan replaceRight(LogicalPlan newRight) {
 
-    protected AttributeSet computeReferences() {
-        // TODO: this needs to be driven by the join config
-        return Expressions.references(output());
```
This made no sense. And calling `output()` here is bad because `references()` is called in the `fieldNames` method - at which point the `output()` method is illegal to use because it requires a resolved plan.

alex-spies commented on x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/analysis/AnalyzerTests.java:
```diff
@@ -1945,9 +1945,10 @@ public void testLookup() {
+                 * If we hadn't it'd be a field and that'd be fine.
                  */
-                .item(containsString("int{f}"))
+                .item(containsString("int{r}"))
```
This test for LOOKUP_:chicken: is only reverted to how it was before we added `LOOKUP JOIN`.

alex-spies committed:
Merge branch 'main' into fix-lookup-join-output

alex-spies committed:
Ignore tests with lookup join on data nodes

elasticsearchmachine commented:
## 💔 Backport failed
| Status | Branch | Result |
|:------:|:------:|:------:|
| ❌ |  8.x  | Commit could not be cherrypicked due to conflicts |

You can use [sqren/backport](https://github.com/sqren/backport) to manually backport by running `backport --upstream elastic/elasticsearch --pr 117639`

alex-spies commented:
## 💚 All backports created successfully

| Status | Branch | Result |
|:------:|:------:|:------|
|✅|8.x|[<img src="https://img.shields.io/github/pulls/detail/state/elastic/elasticsearch/117687">](https://github.com/elastic/elasticsearch/pull/117687)|

### Questions ?
Please refer to the [Backport tool documentation](https://github.com/sqren/backport)

<!--- Backport version: 9.4.3 -->
